qiime2/core/type/template.py

```diff
@@ -0,0 +1,163 @@
+
+    def is_symbol_supertype(self, other):
+        return self.get_name() == other.get_name()
+
+    def update_ast_expr(self, self_expr, ast):
+        self.update_ast(ast)
+
+    def update_ast(self, ast):
+        pass  # default is to do nothing
+
+
+class TypeTemplate(_BaseTemplate):
+    def __new__(cls, *args, _pickle=False, **kwargs):
+        self = super().__new__(cls)
+        if _pickle:
+            return self
+
+        self.__init__(*args, **kwargs)
+        if list(self.get_field_names()):
+            return IncompleteExp(self)
+        else:
+            return TypeExp(self)
+
+    def __getnewargs_ex__(self):
+        return ((), {'_pickle': True})
+
+    def get_field_names_expr(self, expr):
+        return self.get_field_names()
+
+    @abstractmethod
+    def get_field_names(self):
+        raise NotImplementedError
+
+    def validate_fields_expr(self, self_expr, fields_expr):
+        self.validate_field_count(len(fields_expr))
+        for expr, name in itertools.zip_longest(
+                fields_expr, self.get_field_names_expr(self_expr),
+                fillvalue=IntersectionExp()):
+            if expr.template is None:
+                for exp in expr.members:
+                    if exp.template is None:
+                        for ex in exp.members:
+                            self.validate_field(name, ex.template)
+                    else:
+                        self.validate_field(name, exp.template)
+            else:
+                self.validate_field(name, expr.template)
+
+    def validate_field_count(self, count):
+        exp = len(self.get_field_names())
+        if count != exp:
+            raise TypeError("Expected only %r fields, got %r" % (exp, count))
+
+    @abstractmethod
+    def validate_field(self, name, field):
+        raise NotImplementedError
+
+    def validate_predicate_expr(self, self_expr, predicate_expr):
+        if predicate_expr.template is None:
+            for predicate in predicate_expr.members:
+                self.validate_predicate_expr(self_expr, predicate)
+        else:
+            self.validate_predicate(predicate_expr.template)
+
+    @abstractmethod
+    def validate_predicate(self, predicate):
+        raise NotImplementedError
+
+
+class PredicateTemplate(_BaseTemplate):
+    def __new__(cls, *args, _pickle=False, **kwargs):
+        self = super().__new__(cls)
+        if _pickle:
+            return self
+
+        self.__init__(*args, **kwargs)
+        return PredicateExp(self)
+
+    def __getnewargs_ex__(self):
+        return ((), {'_pickle': True})
+
+    @abstractmethod
+    def __hash__(self, other):
+        raise NotImplementedError
+
+    @abstractmethod
+    def is_symbol_subtype(self, other):
+        raise NotImplementedError
+
+    @abstractmethod
+    def is_symbol_supertype(self, other):
+        raise NotImplementedError
+
+    def unpack_union(self):
+        yield
+        return
```

Comment:
whoa.

Reply:
actually I think this is either backwards, or missing `self`....

Reply:
good news, its actually left-over from that time I was teaching `Choices` to self-expand like a union. We can remove this